### PR TITLE
fature/1156_add_metadata_storing_in_mongosink

### PR DESCRIPTION
**Pull request by sabrine2020**
this feature can be configured using: attr_metadata_store in
agent_ngsi.conf file

### CODE REVIEW

**Comment by frbattid** [comment]
Hi @sabrine2020 , it seems you finally implemented issue https://github.com/telefonicaid/fiware-cygnus/issues/1156, nice! :) Just one thing: it seems the modified files have completely re-indented, thus the files appear as completely modified, which avoids reviewing which parts of the code have been added, removed or modified. I must ask you to improve the PR, or even closing it and opening a new one. Thanks!

**Comment by fgalan** [comment]
@sabrine2020 hint: configure your IDE to check style based in this document https://github.com/telefonicaid/fiware-cygnus/blob/master/telefonica_checkstyle.xml. That should ensure that your PR fits into the "official" Cygnus code style ;)

**Comment by sabrine2020** [comment]
I reverted the first commit and I commit a new one with modifications more clear. thanks.

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSink.java [hunk 1]
```diff
@@ -235,11 +247,51 @@ public void aggregate(NGSIEvent cygnusEvent) {
                 
                 LOGGER.debug("[" + getName() + "] Processing context attribute (name=" + attrName + ", type="
                         + attrType + ")");
-                Document doc = createDoc(recvTimeTs, entityId, entityType, attrName, attrType, attrValue);
+                
+                Document doc;
+				if(attrMetadataStore.equals("true")){
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSink.java]: Ident have improved *a lot* :) but seem still to be different from existing one... checking style would be a good idea before merging the PR.

File: cygnus-ngsi/conf/agent_ngsi.conf.template [hunk 2]
```diff
@@ -257,6 +257,8 @@ cygnus-ngsi.sinks.mongo-sink.channel = mongo-channel
 #cygnus-ngsi.sinks.mongo-sink.data_model = dm-by-entity
 # how the attributes are stored, either per row either per column (row, column)
 #cygnus-ngsi.sinks.mongo-sink.attr_persistence = row
+# true if the metadata will be stored, false by default
```
Comment (fgalan) [inline cygnus-ngsi/conf/agent_ngsi.conf.template]: Probably this change adding a new parameter should be mirrored in .md documentation files that cite configuration (is there any of such fragments?) and also explained in accompanying text.

@pratid what do you think?

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSink.java (hunk 1 above)
Comment (sabrine2020) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSink.java]: I used the check style but no error has been showed, may be I miss configured?
checkstyleconfig

Comment (sabrine2020) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSink.java]: checkstyleconfig_1

**Comment by frbattid** [comment]
An entry must be added to CAHNGES_NEXT_RELEASE, somethig like:

    - [cygnus-ngsi][feature] Add metadata storing in NGSIMongoSink (#1156)

File: cygnus-ngsi/conf/agent_ngsi.conf.template (hunk 2 above)
Comment (fgalan) [inline cygnus-ngsi/conf/agent_ngsi.conf.template]: @frbattid what do you think? (sorry @pratid ! ;)

**Comment by sabrine2020** [comment]
Hello, is it ok for this pull to be integrated in the next version? Thanks!

**Comment by fgalan** [comment]
I think that there are yet pending changes at least regarding documentation and CHANGES_NEXT_RELEASE file...

**Comment by sabrine2020** [comment]
How it works? should I do the changes? thanks.

**Comment by frbattid** [comment]
@sabrine2020 , yes, you must edit `CHANGES_NEXT_RELEASE` in order to add the line I suggested you, and you must edit the involved documentation pieces. Thanks!

**Comment by sabrine2020** [comment]
Requested changes done, tell me if it is ok for these modifications to be integrated in the next release.. Thanks!

**Comment by frbattid** [comment]
@sabrine2020 , we are almost done, thanks! Just a couple of things:

* Default value for `cygnus-ngsi.sinks.mongo-sink.attr_metadata_store` is `false` (which is OK for me). Nevertheless, such parameter, when appearing in the template and the documentation files is commented (OK for me again) and set to `true`... I would like it is set to `false`, since traditionally we have put the default value to commented parameters.
* If you have a look on the proposed changes, indentation is not fine as @fgalan detected. I'll try to check your branch in order to see what happens with the style check.

**Comment by sabrine2020** [comment]
@frbattid thanks! I changed the commented parameters value to false in the tempalte. Keep me updated about the style check, I dont know why it is not showing errors in my eclipse. May be I did not well configure it.

**Comment by fgalan** [comment]
>  I changed the commented parameters value to false in the tempalte.

This one seems to remain: https://github.com/telefonicaid/fiware-cygnus/pull/1416/files#diff-7451304b8d110573f430575a4f551d9bR261

**Comment by sabrine2020** [comment]
oh yep sorry, I updated it

**Comment by fgalan** [comment]
LGTM (precluding that the style check that @frbattid will do on the branch is ok). Thanks for your contribution!

**Comment by frbattid** [comment]
Hi @sabrine2020 , regarding indent, I've created the following patch for fixing it:

https://gist.github.com/frbattid/bf55faa296f6a4d97e1f7ef8e7fa5250

You only have to copy the content of the above gist in a `NGSIMongoSink.java.patch` file within the path you have your `NGSIMongoSink.java` file and then run:

$ git apply NGSIMongoSink.java.patch

Once patched, just remove `NGSIMongoSink.java.patch` and upload `NGSIMongoSink.java` as usual in a new commit.

Hope this helps!

**Comment by sabrine2020** [comment]
Hi @frbattid thanks!
I got an error while applying the patch:
```
git apply NGSIMongoSink.java.patch
fatal: corrupt patch at line 120
```
Any idea why?
Best,
Sabrine

**Comment by frbattid** [comment]
Mmm, can you get the MD5 of your `NGSIMongoSink.java.patch` file and check if it matches `282fd42c0f1d45be91b8b54091415b2b`?

**Comment by sabrine2020** [comment]
@frbattid  apparently no:
```
CertUtil -hashfile NGSIMongoSink.java.patch MD5
Hachage MD5 du fichier NGSIMongoSink.java.patch :
6b 23 34 29 b6 1a ed b5 f9 2b d7 65 29 8f 24 b1
```

**Comment by frbattid** [comment]
I downloaded the gist content using `wget` and I was able to reproduce your error. Nevertheless, I added a final new line (i.e. a crriage return at the end of line 120) and it worked. Try doing that, please.

**Comment by sabrine2020** [comment]
@frbattid : Thanks, I executed the patch after the add of a new line, I got no error but I don't see changes in the file NGSIMongoSink.java
```
>git apply NGSIMongoSink.java.patch
>git status
On branch master
Your branch is up-to-date with 'origin/master'.
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        ../../../../../../../../.classpath
        ../../../../../../../../.project
        ../../../../../../../../.settings/
        NGSIMongoSink.java.patch

nothing added to commit but untracked files present (use "git add" to track)
```
weird :s

**Comment by frbattid** [comment]
So... the only explanation I can found is: the number of tabs you have inserted in order to properly indent the code is correct, but then the way your IDE translates a tab character to white spaces is not the same. I mean, my Netbeans configuration says "Number of Spaces per Indent: 4", thus I would try to have the same value in your IDE.

**Comment by frbattid** [comment]
Other possibility is to configure your IDE in order it uses white spaces instead of tabs. I think that's the case of Eclipse: http://stackoverflow.com/questions/407929/how-do-i-change-eclipse-to-use-spaces-instead-of-tabs

**Comment by fgalan** [comment]
It seem that last commit has changed the whole .java file :(. See https://github.com/telefonicaid/fiware-cygnus/pull/1416/files#diff-ef645f2c926758c5b19303a7bc50f80f

I'd suggest to rollback that commit (115c0b6)

**Comment by sabrine2020** [comment]
yep :s, I applied the patch using Eclipse IDE and I changed the directory a and b by the current ./
but it did not work. it updated the whole file :( not only the added lines..

**Comment by fgalan** [comment]
Sometime IDEs are a problem rather than a solution :)

What about applying the .patch with ye olde CLI command? I mean, something like:

```
patch -pX < NGSIMongoSink.java.patch
```

('pX' should actually be 'p0' or 'p1'.... I don't remember the detail)

**Comment by sabrine2020** [comment]
Even after the change of my IDE configuration to space only, Applying the patch using the cmd does not change the java file and trying to update it using eclipse (after changing the patch directory) update the whole file...
Don't know what can be done to fix the indentation :s

**Comment by fgalan** [comment]
...and if everything else fails, what about "mirroring" the changes in the .patch manually on the .java file with a plain editor (e.g. vim)? Look for tabs, Tabs are evil :) https://www.emacswiki.org/emacs/TabsAreEvil

**Comment by sabrine2020** [comment]
I think there is a problem with the index number in the patch, it is not the same as mine.
since I am getting this error: `error: NGSIMongoSink.java: does not exist in index`

```
>git am --signoff < NGSIMongoSink.java.patch
Applying: Signed-off-by: sabrine 
.git/rebase-apply/patch:24: trailing whitespace.
        } // if else
.git/rebase-apply/patch:45: trailing whitespace.

.git/rebase-apply/patch:106: trailing whitespace.
            return doc;
error: NGSIMongoSink.java: does not exist in index
Patch failed at 0001 Signed-off-by: sabrine 
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

In git patch file the index is:
`index 1c43c21..8b5d115 100644`
any idea how to update this value to use my current index? Thanks

**Comment by fgalan** [comment]
What about to apply the patch using plain `patch` command (i.e. without using `git`)?

**Comment by sabrine2020** [comment]
I think the patch command did it, could you check if the indentation is correct? thanks!

**Comment by frbattid** [comment]
Yeah, well done @sabrine2020 !!

**Comment by sabrine2020** [comment]
@frbattid @fgalan : thanks for your support!

**Pull request merged**